Implement serialisation of RTCEncodedAudioFrame and RTCEncodedVideoFrame

youennf opened the pull request:
#### 377e0898e62e6d206ba830bb5904f1bdc9c0defa
<pre>
Implement serialisation of RTCEncodedAudioFrame and RTCEncodedVideoFrame
<a href="https://rdar.apple.com/148244334">rdar://148244334</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290756">https://bugs.webkit.org/show_bug.cgi?id=290756</a>

Reviewed by Eric Carlson.

As per specification, we make RTCEncodedAudioFrame and RTCEncodedVideoFrame serialized.
To do so, we make RTCRtpTransformableFrame thread safe refcounted and cloneable.
When serializing, we take a clone of RTCRtpTransformableFrame and we reconstruct the related RTCEncodedAudioFrame/RTCEncodedVideoFrame from it.
This only implements serialization of RTCEncodedAudioFrame/RTCEncodedVideoFrame within the same process which is the desired scope for now.

Some tests in WPT webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https.html are expected to fail, they will be fixed in a follow-up.

Covered by added WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-drop-frames-worker.js: Added.
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-no-write-worker.js: Added.
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-worker.js: Added.
(appendToBuffer):
(ModifyAndWrite):
(RestoreAndWrite):
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.js: Added.
(onrtctransform.async event):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/WEB_FEATURES.yml: Added.
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/RTCEncodedFrame.cpp:
(WebCore::RTCEncodedFrame::serialize):
* Source/WebCore/Modules/mediastream/RTCEncodedFrame.h:
* Source/WebCore/Modules/mediastream/RTCRtpTransformableFrame.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::clone):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::name):
(WebCore::isTypeExposedToGlobalObject):
(WebCore::CloneSerializer::serializeCryptoKey):
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpRTCEncodedAudioFrame):
(WebCore::CloneSerializer::dumpRTCEncodedVideoFrame):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::takeSerializedRTCEncodedAudioFrames):
(WebCore::CloneDeserializer::takeSerializedRTCEncodedVideoFrames):
(WebCore::CloneDeserializer::readRTCEncodedAudioFrame):
(WebCore::CloneDeserializer::readRTCEncodedVideoFrame):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::validateSerializedResult):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/293185@main">https://commits.webkit.org/293185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc02660e7bddfe7ff19d42100cf9f380ff0de5d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98117 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103233 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48646 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100162 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26199 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74707 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31889 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13670 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88658 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55067 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13454 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6591 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48088 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83431 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6671 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105610 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25203 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18361 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25576 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83155 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21006 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27784 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5466 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18843 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25162 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30336 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28298 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-drop-frames-worker.js:
```diff
@@ -0,0 +1,34 @@
+        if (chunk.done)
+            return;
+        if (transformer.options.name === 'sender') {
+            if(countRead == 5) {
```
Tiny nit: should have a space between `if` and `(`

youennf replied:
Given this is a WPT test, it seems best not to change it here as import time but in a follow-up

eric-carlson commented on LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https-expected.txt:
```diff
@@ -0,0 +1,7 @@
+
+
+FAIL Receiver and sender read, modifiy and write video frames. assert_equals: expected "got expected" but got "unexpected value: lastByte (got 1, expected 1),  frame data length (got 532, expected 0), rtpTimestamp (got undefined, expected null)"
+FAIL Receiver and sender read, modifiy and write audio frames. assert_equals: expected "got expected" but got "unexpected value: lastByte (got 1, expected 1),  frame data length (got 98, expected 0), rtpTimestamp (got undefined, expected null)"
```
If these are expected to fail you might mention it in the commit message

youennf replied:
OK, will do in the commit message.

youennf committed:
Implement serialisation of RTCEncodedAudioFrame and RTCEncodedVideoFrame
rdar://148244334
https://bugs.webkit.org/show_bug.cgi?id=290756

Reviewed by Eric Carlson.

As per specification, we make RTCEncodedAudioFrame and RTCEncodedVideoFrame serialized.
To do so, we make RTCRtpTransformableFrame thread safe refcounted and cloneable.
When serializing, we take a clone of RTCRtpTransformableFrame and we reconstruct the related RTCEncodedAudioFrame/RTCEncodedVideoFrame from it.
This only implements serialization of RTCEncodedAudioFrame/RTCEncodedVideoFrame within the same process which is the desired scope for now.

Some tests in WPT webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https.html are expected to fail, they will be fixed in a follow-up.

Covered by added WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-drop-frames-worker.js: Added.
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-no-write-worker.js: Added.
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform-worker.js: Added.
(appendToBuffer):
(ModifyAndWrite):
(RestoreAndWrite):
(onrtctransform.async process):
(onrtctransform):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-encoded-transform.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/RTCRtpScriptTransform-sender-worker-single-frame.js: Added.
(onrtctransform.async event):
* LayoutTests/imported/w3c/web-platform-tests/webrtc-encoded-transform/WEB_FEATURES.yml: Added.
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/RTCEncodedFrame.cpp:
(WebCore::RTCEncodedFrame::serialize):
* Source/WebCore/Modules/mediastream/RTCEncodedFrame.h:
* Source/WebCore/Modules/mediastream/RTCRtpTransformableFrame.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.cpp:
(WebCore::LibWebRTCRtpTransformableFrame::clone):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpTransformableFrame.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::name):
(WebCore::isTypeExposedToGlobalObject):
(WebCore::CloneSerializer::serializeCryptoKey):
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpRTCEncodedAudioFrame):
(WebCore::CloneSerializer::dumpRTCEncodedVideoFrame):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::deserialize):
(WebCore::CloneDeserializer::CloneDeserializer):
(WebCore::CloneDeserializer::takeSerializedRTCEncodedAudioFrames):
(WebCore::CloneDeserializer::takeSerializedRTCEncodedVideoFrames):
(WebCore::CloneDeserializer::readRTCEncodedAudioFrame):
(WebCore::CloneDeserializer::readRTCEncodedVideoFrame):
(WebCore::CloneDeserializer::readTerminal):
(WebCore::validateSerializedResult):
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::create):
(WebCore::SerializedScriptValue::deserialize):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/293185@main
